### PR TITLE
Create override plugin

### PR DESCRIPTION
**Pull request by fejta**
Allow repo administrators to `/override foo` to mark foo as passed

/assign @cjwagner @stevekuznetsov @BenTheElder 
/lint

### CODE REVIEW

File: prow/github/client.go [hunk 1]
```diff
@@ -578,6 +578,40 @@ func (c *Client) ListOrgMembers(org, role string) ([]TeamMember, error) {
...
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// GetUserPermission returns the user's permission level for a repo
+//
+// https://developer.github.com/v3/repos/collaborators/#review-a-users-permission-level
+func (c *Client) GetUserPermission(org, repo, user string) (string, error) {
```
Comment (stevekuznetsov) [inline prow/github/client.go]: Do we want to `type UserPermission string` the return value here?

File: prow/plugins/override/override.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+func handleGenericComment(pc plugins.PluginClient, e github.GenericCommentEvent) error {
+	return handle(pc.GitHubClient, pc.Logger, &e)
+}
+
+func authorized(gc githubClient, log *logrus.Entry, org, repo, user string) bool {
+	ok, err := gc.HasPermission(org, repo, user, github.RoleAdmin)
+	if err != nil {
+		log.Warnf("cannot determine whether %s is an admin of  %s/%s", user, org, repo)
```
Comment (stevekuznetsov) [inline prow/plugins/override/override.go]: `log.WithError(err).Warnf(...`

Comment (BenTheElder) [reply prow/plugins/override/override.go]: good catch

**Comment by fejta** [comment]
> we will still fail on the ProwJob record for tests we ran

is that a regression?

**Comment by stevekuznetsov** [comment]
No, it's not a regression. `tide` holds `ProwJob` CRD records as the sole source of truth for jobs that ran on `Prow`. If the record in etcd is of a failed test, it does not matter what the GitHub status context reads.

**Comment by stevekuznetsov** [comment]
`/skip` will only update GitHub status contexts for non-blocking jobs, so ones that `tide` would not have cared about anyway

**Comment by fejta** [comment]
So how do we want to handle this? Should I create a fake prow job resource with a passing result?

**Comment by stevekuznetsov** [comment]
Yeah I guess we would need a fake ProwJob CRD record which seems kind of yucky ... @cjwagner and @spxtr can speak to why the use of the CRD is in `tide` -- IIRC originally `tide` did not do _anything_ with contexts and just used the CRD

**Comment by fejta** [comment]
> how do we want to handle this

Okay I plan to make `/override` create a prowjob with a status of success and an agent of override :-/

**Comment by fejta** [comment]
What tide needs:

https://github.com/kubernetes/test-infra/blob/7ae75c69e505148af78fc5c48835c49a5a99a2f8/prow/tide/tide.go#L212
https://github.com/kubernetes/test-infra/blob/7ae75c69e505148af78fc5c48835c49a5a99a2f8/prow/tide/tide.go#L226
https://github.com/kubernetes/test-infra/blob/7ae75c69e505148af78fc5c48835c49a5a99a2f8/prow/tide/tide.go#L244
https://github.com/kubernetes/test-infra/blob/7ae75c69e505148af78fc5c48835c49a5a99a2f8/prow/tide/tide.go#L726
https://github.com/kubernetes/test-infra/blob/7ae75c69e505148af78fc5c48835c49a5a99a2f8/prow/tide/tide.go#L438-L457

How we create jobs:
https://github.com/kubernetes/test-infra/blob/7ae75c69e505148af78fc5c48835c49a5a99a2f8/prow/plugins/trigger/trigger.go#L272-L292

File: prow/github/client.go (hunk 1 above)
Comment (fejta) [reply prow/github/client.go]: Done

File: prow/plugins/override/override.go (hunk 2 above)
Comment (fejta) [reply prow/plugins/override/override.go]: Done

**Comment by fejta** [comment]
PTAL

File: prow/plugins/override/override.go [hunk 3]
```diff
@@ -129,19 +190,44 @@ func handle(gc githubClient, log *logrus.Entry, e *github.GenericCommentEvent) e
...
 		log.Info(msg)
-		gc.CreateComment(org, repo, number, plugins.FormatResponseRaw(e.Body, e.HTMLURL, user, msg))
+		oc.CreateComment(org, repo, number, plugins.FormatResponseRaw(e.Body, e.HTMLURL, user, msg))
 	}()
 
 	for _, status := range statuses {
 		if status.State == github.StatusSuccess || !overrides.Has(status.Context) {
 			continue
 		}
+		// First create the overriddent prow result if necessary
```
Comment (stevekuznetsov) [inline prow/plugins/override/override.go]: `s/overriddent/overridden/`

Comment (fejta) [reply prow/plugins/override/override.go]: Done

**Comment by fejta** [comment]
/meow space

Woohoo! Thanks :D

**Pull request merged**